std/digest/murmurhash3.d

```diff
@@ -0,0 +1,734 @@
+/**
+Computes MurmurHash3 hashes of arbitrary data. MurmurHash is a non-cryptographic
+hash function suitable for general hash-based lookup.
+
+MurmurHash3 yields a 32-bit or 128-bit hash value.
+
+This module conforms to the APIs defined in $(D std.digest.digest).
+To understand the differences between the template and the OOP API, see $(D std.digest.digest).
+
+This module publicly imports $(D std.digest.digest) and can be used as a stand-alone module.
+
+License:   $(WEB www.boost.org/LICENSE_1_0.txt, Boost License 1.0).
+CTFE: Digests do not work in CTFE
+Authors: Guillaume Chatelet $(BR)
+References:
+  $(LINK2 https://code.google.com/p/smhasher/wiki/MurmurHash3, reference implementation)
+  $(LINK2 https://en.wikipedia.org/wiki/MurmurHash, Wikipedia on MurmurHash)
+*/
+module std.digest.murmurhash3;
+
+public import std.digest.digest;
+
+@safe:
+
+alias MurmurHash3_x86_32 = Piecewise!SMurmurHash3_x86_32;
+alias MurmurHash3_x86_128 = Piecewise!SMurmurHash3_x86_128;
+alias MurmurHash3_x64_128 = Piecewise!SMurmurHash3_x64_128;
+
+alias MurmurHash3_x86_32Digest = WrapperDigest!MurmurHash3_x86_32;
+alias MurmurHash3_x86_128Digest = WrapperDigest!MurmurHash3_x86_128;
+alias MurmurHash3_x64_128Digest = WrapperDigest!MurmurHash3_x64_128;
+
+/**
+MurmurHash3 cannot put chunks smaller than blockSizeInBytes at a time. This
+struct stores remainder bytes in a buffer and pushes it as soon as the  block is
+complete or during finalization.
+*/
+struct Piecewise(Hasher)
+{
+    alias Block = Hasher.Block;
+    enum blockSizeInBytes = Block.sizeof;
+    enum blockSize = bits!Block;
+
+    Hasher hasher;
+    ubyte[blockSizeInBytes] buffer;
+    size_t bufferSize = 0;
+
+    void start()
+    {
+        this = Piecewise.init;
+    }
+
+    void put(scope const(ubyte)[] data...) pure nothrow
+    {
+        // Buffer should never be full while entring this function.
+        assert(bufferSize < blockSizeInBytes);
+        // Complete left over from last call if any.
+        if (bufferSize > 0)
+        {
+            import std.algorithm.comparison : min;
+
+            immutable available = blockSizeInBytes - bufferSize;
+            immutable copySize = min(available, data.length);
+            buffer[bufferSize .. bufferSize + copySize] = data[0 .. copySize];
+            data = data[copySize .. $];
+            bufferSize += copySize;
+        }
+        // Push left over if it reached the size of a block.
+        if (bufferSize == blockSizeInBytes)
+        {
+            hasher.putBlocks(cast(Block[1]) buffer);
+            bufferSize = 0;
+        }
+        // Pushing as many consecutive blocks as possible.
+        immutable consecutiveBlocks = alignDownTo(data.length);
+        hasher.putBlocks(cast(const(Block)[]) data[0 .. consecutiveBlocks]);
```

Comment:
Is it safe? How it would work if data is not aligned? (may be create a small scope buffer for tail and and unaligned head?)